geometry3d/basicMorphoFilter.cpp

```diff
@@ -0,0 +1,206 @@
+/**
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+/**
+ * @file basicMorphoFilter.cpp
+
+ * @author Bertrand Kerautret (\c kerautre@loria.fr)
+ * LORIA (CNRS, UMR 7503), University of Nancy, France
+ *
+ * @date 2015/28/11
+ *
+ */
+
+
+#include <iostream>
+#include <fstream>
+
+#include "DGtal/base/Common.h"
+#include "DGtal/helpers/StdDefs.h"
+#include "DGtal/io/readers/VolReader.h"
+#include "DGtal/io/writers/GenericWriter.h"
+
+#include "DGtal/images/ImageContainerBySTLVector.h"
+#include <boost/program_options/options_description.hpp>
+#include <boost/program_options/parsers.hpp>
+#include <boost/program_options/variables_map.hpp>
+
+using namespace DGtal;
+namespace po = boost::program_options;
+
+typedef typename Z3i::Point Point3D;
+typedef ImageContainerBySTLVector<Z3i::Domain, unsigned char> Image3D;
+
+
+// size= width/2
+template<typename TImage>
+TImage
+applyMedian(const TImage &anImage, unsigned int size){
+  unsigned int imageSize = (anImage.domain().upperBound()[0]-anImage.domain().lowerBound()[0])*
+    (anImage.domain().upperBound()[1]-anImage.domain().lowerBound()[1])*
+    (anImage.domain().upperBound()[2]-anImage.domain().lowerBound()[2]);
+  
+  trace.progressBar(0, imageSize);
+  typedef typename TImage::Domain::ConstIterator ImageDomIterator;
+  TImage imageRes (anImage.domain());
+  unsigned int cpt = 0;
+  
+  for(ImageDomIterator it = anImage.domain().begin(); it != anImage.domain().end(); ++it){
+    trace.progressBar(cpt, imageSize);
+    Point3D pt = *it;
+    std::vector<unsigned char> vectVal;
+    for (int k = -(int)size; k <= (int) size; k++) {
+      for (int l = -(int)size; l <= (int) size; l++) {
+        for (int m = -(int)size; m <= (int) size; m++) {
+          Point3D p (((int)(pt[0]))+k, (int)(pt[1])+l, ((int)(pt[2]))+m);
```

Comment:
ok for the DT but could you please just add a test here to make sure that p is in the ball ?
e.g. ``` (p-pt).norm()<size``` 

I think it is important to not have squares structural elements..